File: src/main/scala/org/broadinstitute/hail/utils/Logging.scala
package org.broadinstitute.hail.utils

import org.slf4j.{Logger, LoggerFactory}

trait Logging {
  @transient var log_ : Logger = _

  def log: Logger = {
    if (log_ == null)
      log_ = LoggerFactory.getLogger("Hail")
    log_
  }

  def info(msg: String) {
    log.info(msg)
    System.err.println("hail: info: " + msg)
  }

  def warn(msg: String) {
    log.warn(msg)
    System.err.println("hail: warning: " + msg)
  }

  def error(msg: String) {
    log.error(msg)
    System.err.println("hail: error: " + msg)
  }
}

File: hail/src/main/c/Logging.cpp
#include "hail/Logging.h"
#include "hail/Upcalls.h"
#include <cstdio>
#include <string>

namespace hail {

void set_test_msg(const char* msg) {
  UpcallEnv e;
  e.set_test_msg(msg);
}

void info(const char* msg) {
  UpcallEnv e;
  e.info(msg);
}

void warn(const char* msg) {
  UpcallEnv e;
  e.warn(msg);
}

void error(const char* msg) {
  UpcallEnv e;
  e.error(msg);
}

}
